### PR TITLE
add aws-node-termination-handler add-on

### PR DESCRIPTION
**Pull request by askulkarni2**
*Issue #, if available:* #209

*Description of changes:*
Add AWS Node Termination Handler Add-On that can be configured in IMDS mode (default) or Queue mode.

Unit Tests:
```
> @aws-quickstart/ssp-amazon-eks@0.12.5 test
> jest --verbose

  console.log
    SSP tracking (qs-1s1r465hk)

      at Object.<anonymous> (test/stacks.test.ts:19:17)

  console.log
    My awesome description SSP tracking (qs-1s1r465hk)

      at Object.<anonymous> (test/stacks.test.ts:27:17)

  console.log
    SSP tracking (qs-1s1r465k6)

      at Object.<anonymous> (test/stacks.test.ts:135:17)

 PASS  test/stacks.test.ts (15.195 s)
  ✓ Named resource providers are correctly registered and discovered (222 ms)
  Unit tests for EKS Blueprint
    ✓ Usage tracking created (436 ms)
    ✓ Stack creation fails due to missing add-on dependency (194 ms)
    ✓ Stack creation fails due to wrong node group type for NTH addon (187 ms)
    ✓ Blueprint builder creates correct stack (2821 ms)
    ✓ Pipeline Builder Creates correct pipeline (1067 ms)
    ✓ Nested stack add-on creates correct nested stack (223 ms)

Test Suites: 1 passed, 1 total
Tests:       7 passed, 7 total
Snapshots:   0 total
Time:        15.337 s, estimated 16 s
Ran all test suites.
```

IMDS

```
$kubectl logs ssp-addon-aws-node-termination-handler-cr46q -n kube-system
...
2022/01/12 06:48:25 INF Kubernetes AWS Node Termination Handler has started successfully!
2022/01/12 06:48:25 INF Started watching for event cancellations
2022/01/12 06:48:25 INF Started monitoring for events event_type=SPOT_ITN
2022/01/12 06:48:25 INF Started monitoring for events event_type=SCHEDULED_EVENT
2022/01/12 06:48:25 INF Started monitoring for events event_type=REBALANCE_RECOMMENDATION
```

QUEUE
```
...
2022/01/12 07:49:59 INF Started watching for interruption events
2022/01/12 07:49:59 INF Kubernetes AWS Node Termination Handler has started successfully!
2022/01/12 07:49:59 INF Started watching for event cancellations
2022/01/12 07:49:59 INF Started monitoring for events event_type=SQS_TERMINATE
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by shapirov103 — COMMENTED** [review]
Implementation looks nice! 
a couple of questions/comments

File: docs/addons/aws-node-termination-handler.md [hunk 1]
```diff
@@ -0,0 +1,40 @@
+# AWS Node Termination Handler
+
+This project ensures that the Kubernetes control plane responds appropriately to events that can cause your EC2 instance to become unavailable, such as EC2 maintenance events, EC2 Spot interruptions, ASG Scale-In, ASG AZ Rebalance, and EC2 Instance Termination via the API or Console. If not handled, your application code may not stop gracefully, take longer to recover full availability, or accidentally schedule work to nodes that are going down. For more information see [README.md][https://github.com/aws/aws-node-termination-handler#readme].
```
[the remark this answers is not in the document]
Comment (askulkarni2) [reply docs/addons/aws-node-termination-handler.md:3]: This [issue](https://github.com/aws/aws-node-termination-handler/issues/186) and [EKS Workshop](https://www.eksworkshop.com/beginner/150_spotnodegroups/spotlifecycle/) has description on why NTH is not needed for MNG and MNG/Spot. I updated the doc with references to  both. And note for Fargate as well.

File: lib/cluster-providers/asg-cluster-provider.ts [hunk 2]
```diff
@@ -87,7 +87,7 @@ export class AsgClusterProvider implements ClusterProvider {
         const outputClusterName = true
         const version = this.props.version
         const privateCluster = this.props.privateCluster ?? valueFromContext(scope, constants.PRIVATE_CLUSTER, false);
-        const vpcSubnets = (privateCluster === true) ? [{ subnetType: ec2.SubnetType.PRIVATE }] : this.props.vpcSubnets;
+        const vpcSubnets = (privateCluster === true) ? [{ subnetType: ec2.SubnetType.PRIVATE_WITH_NAT }] : this.props.vpcSubnets;
```
[the remark this answers is not in the document]
Comment (askulkarni2) [reply lib/cluster-providers/asg-cluster-provider.ts:90]: `PRIVATE` and `ISOLATED` are now deprecated. New enums are `PUBLIC`, `PRIVATE_WITH_NAT` and `PRIVATE_ISOLATED`

File: lib/addons/aws-node-termination-handler/index.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+      // Deploy the helm chart
+      const awsNodeTerminationHandlerChart = this.addHelmChart(clusterInfo, {
+        enableSqsTerminationDraining: true,
+        queueURL: queue.queueUrl,
+        serviceAccount: {
+          create: false,
+          name: serviceAccount.serviceAccountName,
+        },
+        enablePrometheusServer: true
```
[the remark this answers is not in the document]
Comment (askulkarni2) [reply lib/addons/aws-node-termination-handler/index.ts:113]: removed

File: docs/addons/aws-node-termination-handler.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+The aws-node-termination-handler (NTH) can operate in two different modes: Instance Metadata Service (IMDS) or the Queue Processor. In the SSP, we provision the NTH in Queue Processor mode. This means that NTH will monitor an SQS queue of events from Amazon EventBridge for ASG lifecycle events, EC2 status change events, Spot Interruption Termination Notice events, and Spot Rebalance Recommendation events. When NTH detects an instance is going down, NTH uses the Kubernetes API to cordon the node to ensure no new work is scheduled there, then drain it, removing any existing work.
```
[the remark this answers is not in the document]
Comment (askulkarni2) [reply docs/addons/aws-node-termination-handler.md:5]: Queue processor handles EC2 Auto-Scaling Group Termination Lifecycle Hooks to take care of ASG Scale-In, AZ-Rebalance, Unhealthy Instances, and EC2 Status Change Events which is not a feature in the IMDS processor. See https://github.com/aws/aws-node-termination-handler#which-one-should-i-use. 

But I think we should give the customer a choice and default to IMDS to keep it simple.

**Comment by shapirov103** [comment]
@askulkarni2  could you rebase when you get a chance?

File: lib/utils/cluster-utils.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    action: 'createOrUpdateTags',
+    parameters: {
+      Tags: tagList
+    },
+    physicalResourceId: customResource.PhysicalResourceId.of(
+      `${autoScalingGroup}-asg-tag`
+    )
+  };
+
+  new customResource.AwsCustomResource(scope, 'asg-tag', {
```
Comment (shapirov103) [inline lib/utils/cluster-utils.ts:44]: No native CDK way to tag ASG other then custom resource? is there an issue on it?

Comment (askulkarni2) [reply lib/utils/cluster-utils.ts:44]: There doesn't seem to be a way to tag resources after they have been created. I saw that @youngjeong46 added a util to tag subnets. We could create a separate construct for the `custom-resource` so it only provisions one resource. I fast follow that as a separate issue.

**Comment by askulkarni2** [comment]
> @askulkarni2 could you rebase when you get a chance?

Done, re-rerunning e2e tests.

File: lib/utils/cluster-utils.ts (hunk 5 above)
Comment (askulkarni2) [reply lib/utils/cluster-utils.ts:44]: https://github.com/aws-quickstart/ssp-amazon-eks/issues/272

**Review by shapirov103 — APPROVED** [review]
LGTM

**Pull request merged**